### DISCUSSION TITLE
Atomic writes in ConcurrentDictionary

### DISCUSSION THREAD

**Discussion by sakno** [dotnet/runtime#65600] [Q&A] [answered]
Source code of `ConcurrentDictionary` contains the following comment:
https://github.com/dotnet/runtime/blob/3ae87395f638a533f37b8e3385f6d3f199a72f4f/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L63-L64

But the current revision of ECMA-335 standard says:
> A conforming CLI shall guarantee that read and write access to properly aligned memory
locations no larger than the native word size

Value in `Node` is always aligned to machine word. I'm supposing that calculation of `s_isValueWriteAtomic` can be replaced with a single line of code:
```csharp
s_isValueWriteAtomic = Unsafe.SizeOf<TValue>() <= IntPtr.Size;
```

Currently, the type like `struct SmallStruct { int X, Y }` is not considered by the current implementation as tearing-free type on 64-bit platform.

I checked my assumption using SharpLab:
```csharp
public class C {
    public struct S
    {
        public int X, Y;
    }
    
    public static void M(in S x1, out S x2) => x2 = x1;
    
    public static void M2(in long x1, out long x2) => x2 = x1;
}
```

Assembly code:
```asm
; Core CLR 6.0.121.56705 on amd64

C..ctor()
    L0000: ret

C.M(S ByRef, S ByRef)
    L0000: mov rax, [rcx]
    L0003: mov [rdx], rax
    L0006: ret

C.M2(Int64 ByRef, Int64 ByRef)
    L0000: mov rax, [rcx]
    L0003: mov [rdx], rax
    L0006: ret
```

As we can see here, read/write of word-sized value type are atomic.

**EDIT**: Any type that is less than or equal to `native int` is eligible for atomic write without risk of tearing.

**Comment by danmoseley** [comment, marked as answer]
That seems reasonable to me, if you want to throw up a PR maybe we will discover whether it's not 🙂

**Reply by sakno** [reply]
Sure, I can make a PR. Also I'll check other places across .NET codebase. Probably, `ConcurrentDictionary` is not the only place where the _atomicity_ of arbitrary type is not accurate.
